src/org/labkey/test/components/ui/grids/CustomizeGridDialog.java

```diff
@@ -0,0 +1,491 @@
+        elementCache().undoEditsButton.click();
+        return this;
+    }
+
+    /**
+     * Is the 'Undo Edits' button enabled.
+     *
+     * @return True if enabled false otherwise.
+     */
+    public boolean isUndoEditsEnabled()
+    {
+        // It looks like .isEnabled is not accurate if disabled-action-text class attribute is used to disable a tag.
+
+        return !elementCache().undoEditsButton
+                .getAttribute("class").toLowerCase()
+                .contains("disabled-action-text");
+    }
+
+    private List<WebElement> getShownInGridListItems(String fieldLabel)
+    {
+        List<WebElement> listItems = elementCache().getListItemElements(elementCache().shownInGridPanel(), fieldLabel);
+
+        Assert.assertFalse(String.format(FIELD_NOT_IN_GRID, fieldLabel),
+                listItems.isEmpty());
+
+        return listItems;
+    }
+
+    /**
+     * Click the 'Update Grid' button and wait for the grid to update and the dialog to close before returning.
+     */
+    public void clickUpdateGrid()
+    {
+        linkedComponent.doAndWaitForUpdate(() -> {
+            elementCache().updateGridButton.click();
+            waitForClose();
+        });
+    }
+
+    /**
+     * Is the 'Update Grid' button enabled.
+     *
+     * @return True if enabled false otherwise.
+     */
+    public boolean isUpdateGridEnabled()
+    {
+        return elementCache().updateGridButton.isEnabled();
+    }
+
+    @Override
+    protected ElementCache elementCache()
+    {
+        return (ElementCache) super.elementCache();
+    }
+
+    @Override
+    protected ElementCache newElementCache()
+    {
+        return new ElementCache();
+    }
+
+    protected class ElementCache extends ModalDialog.ElementCache
+    {
+        protected final Locator contentPanelLocator = Locator.byClass("field-modal__col-content");
+
+        // This is present to items in both panels.
+        protected final Locator listItemName = Locator.tagWithClass("div", "field-name");
+
+        protected final WebElement undoEditsButton = Locator.tagWithText("span", "Undo edits")
+                .refindWhenNeeded(this);
+
+        // The checkbox has no id or name.
+        protected final Checkbox checkbox = new Checkbox(Locator.tagWithAttribute("input", "type", "checkbox")
+                .findWhenNeeded(this));
+
+        protected final WebElement updateGridButton = Locator.button("Update Grid")
+                .findWhenNeeded(this);
+
+        // The 'pencil' to edit a label. Only in the Shown in Grid panel.
+        protected final WebElement fieldLabelEdit = Locator.tagWithClass("input", "form-control")
+                .refindWhenNeeded(shownInGridPanel());
+
+        protected WebElement availableFieldsPanel()
+        {
+            return contentPanelLocator.findElements(this).get(0);
+        }
+
+        protected WebElement shownInGridPanel()
+        {
+            return contentPanelLocator.findElements(this).get(1);
+        }
```

Comment:
Are these sub-panels really that unstable that we can't cache them?
```suggestion
        protected final WebElement availableFieldsPanel = contentPanelLocator.index(0).findWhenNeeded(this);
        protected final WebElement shownInGridPanel = contentPanelLocator.index(1).findWhenNeeded(this);
```

Reply:
No they should not be. I think this is a left over from a more generic approach I was thinking of taking but proved to be more complex that what was needed.